flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/DataTypeUtils.java

```diff
@@ -69,11 +71,22 @@ public static DataType transform(DataType typeToTransform, TypeTransformation...
 		return newType;
 	}
 
+	/**
+	 * Checks if the given type is a composite type.
+	 *
+	 * @param dataType Data type to check
+	 * @return True if the type is composite type. True also for legacy composite type.
+	 */
+	public static boolean isCompositeType(DataType dataType) {
+		return dataType instanceof FieldsDataType ||
```

Comment:
We should not check base on the instance of DataType. But use the logical type root/family instead.